### DISCUSSION TITLE
Policy N Times Usage

### DISCUSSION THREAD

**Discussion by darxsky** [eclipse-edc/Connector#4032] [Q&A] [answered]
Good afternoon,
I kindly ask if anyone is willing to clarify this for me. I'm adding the extension that evaluates the number of accesses to a specific asset. Does number of accesses mean the number of negotiations for that asset or the number of transfers? 
Also, I was thinking about how to implement the evaluation function. Is there a way to retrieve, from the negotiation request (/v2/contractnegotiations), the "assetId" and "connectorId" values from the request body (json)? Or are there other ways, exposed by the connector, to retrieve these fields from incoming messages between the two participants? I tried to retrieve the values via getContextData(), but obviously the evaluation function is done before transferring the messages. 
Thank you.

**Comment by darxsky** [comment]
Any ideas?

**Comment by ndr-brt** [comment, marked as answer]
> I'm adding the extension that evaluates the number of accesses to a specific asset. Does number of accesses mean the number of negotiations for that asset or the number of transfers?

what do you mean with "number of accesses"? the EDC domain model never refers to that concept, so it's totally up to you specify what does it mean.

> Also, I was thinking about how to implement the evaluation function. Is there a way to retrieve, from the negotiation request (/v2/contractnegotiations), the "assetId" and "connectorId" values from the request body (json)? Or are there other ways, exposed by the connector, to retrieve these fields from incoming messages between the two participants? I tried to retrieve the values via getContextData(), but obviously the evaluation function is done before transferring the messages.

The "evaluation function" contexts strictly depends on the policy scope, so, for example, in the `transfer.process` context (the one that's evaluated when you start a new transfer process, the policy context will contain the participant agent, the contract agreement and the instant object. Unfortunately these scopes are not documented, but you can find them in the code annotated with the `@PolicyScope` annotation and see how they're used.

**Reply by jimmarino** [reply]
Here's some info:

https://github.com/eclipse-edc/Connector/tree/main/docs/developer/decision-records/2022-03-15-policy-scopes

**Reply by darxsky** [reply]
> The "evaluation function" contexts strictly depends on the policy scope, so, for example, in the transfer.process context (the one that's evaluated when you start a new transfer process, the policy context will contain the participant agent, the contract agreement and the instant object. Unfortunately these scopes are not documented, but you can find them in the code annotated with the @PolicyScope annotation and see how they're used.

 This really helped me a lot. I didn't realize that the policy context depended on the policy scope.

Thanks for your answers.
